presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,203 @@ private boolean advanceToNextRowGroup()
+        return readStruct(type, path, new IntArrayList(), new IntArrayList());
     }
 
-    private Block readStruct(Type type, List<String> path, IntList elementOffsets)
+    private Block readStruct(Type type, List<String> path, IntList definitionLevels, IntList repetitionLevels)
             throws IOException
     {
-        List<TypeSignatureParameter> parameters = type.getTypeSignature().getParameters();
-        Block[] blocks = new Block[parameters.size()];
-        for (int i = 0; i < parameters.size(); i++) {
-            NamedTypeSignature namedTypeSignature = parameters.get(i).getNamedTypeSignature();
+        List<TypeSignatureParameter> fields = type.getTypeSignature().getParameters();
+        Block[] blocks = new Block[fields.size()];
+        IntList fieldDefinitionLevels = null;
+        IntList fieldRepetitionLevels = null;
+        for (int i = 0; i < fields.size(); i++) {
+            fieldDefinitionLevels = new IntArrayList();
+            fieldRepetitionLevels = new IntArrayList();
+            NamedTypeSignature namedTypeSignature = fields.get(i).getNamedTypeSignature();
             Type fieldType = typeManager.getType(namedTypeSignature.getTypeSignature());
             String name = namedTypeSignature.getName();
-            blocks[i] = readBlock(name, fieldType, path, new IntArrayList());
+            blocks[i] = readBlock(name, fieldType, path, fieldDefinitionLevels, fieldRepetitionLevels);
+        }
+        BooleanList structIsNull = new BooleanArrayList();
+        IntList structOffsets = new IntArrayList();
+        calculateStructOffsets(toArray(path), structOffsets, structIsNull, definitionLevels, repetitionLevels, fieldDefinitionLevels, fieldRepetitionLevels);
+        return new RowBlock(0, structIsNull.size(), structIsNull.toBooleanArray(), structOffsets.toIntArray(), blocks);
+    }
+
+    /**
+     * Each struct has three variants of presence:
+     * 1) Struct is not defined, because one of it's optional parent fields is null
+     * 2) Struct is null
+     * 3) Struct is defined and not empty. In this case offset value is increased by one.
+     * <p>
+     * One of the struct's field repetition/definition levels are used to calculate offsets of the struct.
+     */
+    private void calculateStructOffsets(String[] path, IntList structOffsets, BooleanList structIsNull, IntList definitionLevels, IntList repetitionLevels, IntList fieldDefinitionLevels, IntList fieldRepetitionLevels)
+    {
+        int offset = 0;
+        structOffsets.add(offset);
+        if (fieldDefinitionLevels != null) {
+            int maxDefinitionLevel = getMaxDefinitionLevel(path);
+            int maxRepetitionLevel = getMaxRepetitionLevel(path);
+            for (int i = 0; i < fieldDefinitionLevels.size(); i++) {
+                if (fieldRepetitionLevels.get(i) <= maxRepetitionLevel) {
+                    if (isNullField(isRequired(path), fieldDefinitionLevels.get(i), maxDefinitionLevel)) {
```

Comment:
does this `if` handle cases 1 and 2 above in the javadoc?

Reply:
@nezihyigitbasi Second one only (Struct is null). No need to add any information about undefined structs to the block with structs. Such case is possible only when struct is inside some other complex type (nested into another struct, an array element or a map value). When struct is undefined, it's parent is null. So parent block will contain information about being null. Later, when value will be read from the block, it will be decoded starting from the very top level of nesting. As soon as null value is met, nested blocks are not read (value is null, no need to check it's nested fields)